test/integration/server_response/test_error.py

```diff
@@ -0,0 +1,41 @@
+# This module is for requests which result in errors
+
+import requests
+import os
+import stat
+
+def test_file_not_found():
+    print("Request for /nonexistent.html")
+    response = requests.get("http://localhost:8080/nonexistent.html")
+    assert response.status_code == 404
+
+def test_permission_denied_for_file():
```

Comment:
No it works on my machine™ 🥹 

Test runs successful. Server writes: [ERROR] std::ifstream: Permission denied

If I `chmod 000 html/index.html` and open with browser (Chrome) I get 
- 403 for index.html
- 200 for favicon.ico

However if I run `make coverage2` it sometimes passes, and sometimes not. But then several tests don't pass (not only one). I think it's because it reuses the connected socket while running the tests (even though we don't have a session defined..?). If the connection socket is still filled with old bytes, we never receive a "correct" request which leads to a timeout of python request resulting in error.

I added fix where we close connection in case of 
- 413 (too much body bytes), 
- 431 (too much header bytes) and 
- 401 (method not implemented).

I think the tests are also a bit too finicky right now. For example in test_permission_denied_for_file(): if `response = requests.get(..)` throws (e.g. timeout) we never reach `os.chmod(...)` which means the file has incorrect permissions! 
Maybe just a try-except block will do the trick.
Also in your case it could be that with `make coverage2` python requests reaches timeout more often, since server startup also takes longer. We also could just set timeout to a higher value everywhere.

Reply:
Okay, so it's a an error on my side :sweat_smile:
I did not test with ``make coverage2`` but instead performed the tests described bellow.
My setup is also described to make it easier to recognize potential errors.

#  Setup

- I develop on a Mac Mini within our devcontainer. 
- The current branch is: ``12-file-with-no-permission-should-return-403``.
- The devcontainer was rebuilt without cache. 

# Tests

## Automated tests

- ``pytest test/integration/server_response/test_error.py``
- ``pytest test/integration/server_response/test_error.py::test_permission_denied_for_file``

### Outcome

Status Code: 200

## Manual tests

Changed permissions of index.html with this command:
```bash
chmod 000 html/index.html
```

Output of the permissions of index.html:
```bash
---------- 1 vscode vscode 1459 Dec 30 10:39 html/index.html
```

- Accessed URL: ``localhost:8080/index.html`` with Brave inside a private window
- Accessed URL: ``localhost:8080/index.html`` with Chrom inside a private window
- Accessed URL: ``localhost:8080/index.html`` with Brave inside a private window

### Outcome

Status Code: 200

Reply:
It's: ``vscode``

Reply:
Since you are vscode inside dev container, and you start the server manually it shouldn't be able to open the index file..? Can you open it in vscode after chmod 000? I get 
"The editor could not be opened due to an unexpected error: [...]  (NoPermissions (FileSystemError) [...]"
if I try. Close it if you have it already open

Reply:
I can open the file inside the dev container with vscode without any problems

Reply:
Had troubles to test permission removal on MacOS inside the dev container. 

Switched to a Ubuntu VM which I connected to by using VSCode. Then running the test inside the dev container was successfull.